### PR TITLE
Drop NumPy<1.15 support

### LINKED ISSUE #2900: Drop support of older NumPy (<=1.14)?
According to [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html), an unusual NumPy Enhancement Proposal that declares a community-wide policy instead of merely proposing changes to NumPy itself, the support of NumPy <=1.14 will be dropped in early January, 2020, which is a few days later:
> Drop Schedule
> ...
> On Jan 07, 2020 drop support for Numpy 1.14 (initially released on Jan 06, 2018)

Would CuPy consider following NEP 29 so that some test codes can be simplified without worrying too much about backward compatibilities? I've seen this caused hard time for a few PRs.

### PR DESCRIPTION
**Pull request by niboshi**
Following [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html).

Fixes #2900.

### CODE REVIEW

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Fixed test failures and updated `setup.py` as well.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
@niboshi As part of NEP 29 compliance effort, would you also like to drop Python 3.5? This involves  removing a few lines in `setup.py` and the env variable `CUPY_PYTHON_350_FORCE`, as well as revising the installation guidance.

**Comment by niboshi** [comment]
@leofang Thank you, that sounds good as well! I made an issue for that (#2993).

File: tests/cupy_tests/core_tests/test_ndarray_unary_op.py [hunk 1]
```diff
@@ -93,7 +93,6 @@ def test_pos_zerodim(self):
     def test_abs_zerodim(self):
         self.check_zerodim_op_full(operator.abs)
 
-    @testing.with_requires('numpy<1.16')
```
Comment (niboshi) [inline tests/cupy_tests/core_tests/test_ndarray_unary_op.py]: Test passes without this one with newer version of NumPy.

**Comment by niboshi** [comment]
Jenkins failue on `test_dims.py` is because of NumPy 1.18 (being fixed in #2992).
Jenkins shouldn't have installed NumPy 1.18 (chainer-test is configured to install `numpy<1.10` beforehand), but because this PR changes the requirement to `numpy>=1.15`, numpy is reinstalled to the latest (1.18.1), causing this error.

**Comment by emcastillo** [comment]
I will review #2992 and get back to this once merged then, (unless jenkins numpy versioning is fixed first)

**Comment by emcastillo** [comment]
#2992 was merged, but there are conflicts now 😥

**Comment by niboshi** [comment]
Resolved.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
For chainer failure, please look at https://github.com/chainer/chainer-test/pull/556.

File: tests/cupy_tests/padding_tests/test_pad.py [hunk 2]
```diff
@@ -244,7 +240,7 @@ def test_pad_via_func_returning_vector(self, xp, dtype):
         def _padwithtens(vector, pad_width, iaxis, kwargs):
             vector[:pad_width[0]] = 10
             vector[-pad_width[1]:] = 10
-            return vector  # returning vector required by old NumPy (<=1.12)
+            return vector
```
Comment (toslunar) [inline tests/cupy_tests/padding_tests/test_pad.py]: From the comment, I guess the test passes without returning `vector`.

Comment (niboshi) [reply tests/cupy_tests/padding_tests/test_pad.py]: Yes it's tested in `test_pad_via_func_modifying_inplace` above.

I thought this test is written specifically for the case `vector` is returned.
But in the NumPy `pad` docs, no return value is mentioned at all.
I think this test can be removed.

Comment (niboshi) [reply tests/cupy_tests/padding_tests/test_pad.py]: Removed in b0c7be9be52628d41285c2fb8dbf7b7b125beb6e.
Thanks!

**Comment by niboshi** [comment]
Btw I don't understand these errors:
```
17:14:07 FAIL tests/cupyx_tests/scipy_tests/fft_tests/test_fft.py::TestRfftn_param_11_{axes=(-1, -2, -3), norm=None, s=None, shape=(2, 3, 4)}::test_irfftn
17:14:07 FAIL tests/cupyx_tests/scipy_tests/fft_tests/test_fft.py::TestRfftn_param_11_{axes=(-1, -2, -3), norm=None, s=None, shape=(2, 3, 4)}::test_irfftn_overwrite
```
```
17:14:07 E               AssertionError: 
17:14:07 E               Not equal to tolerance rtol=0.0001, atol=1e-07
17:14:07 E               
17:14:07 E               Mismatched elements: 20 / 24 (83.3%)
17:14:07 E               Max absolute difference: 1.03293281
17:14:07 E               Max relative difference: 13.75302148
17:14:07 E                x: array([[[ 6.388732,  0.434309,  0.257819,  0.974329],
17:14:07 E                       [-0.512607, -0.774623, -0.018138,  0.123988],
17:14:07 E                       [-0.762056,  0.123988,  0.231312, -0.774623]],...
17:14:07 E                y: array([[[ 6.104534,  0.15011 , -0.566401,  0.15011 ],
17:14:07 E                       [-0.512607,  0.25831 , -0.018138,  0.145404],
17:14:07 E                       [-0.512607,  0.145404, -0.018138,  0.25831 ]],...
```
Is this numerical instability? :thinking: 
In this PR `@testing.with_requires('numpy>=1.10.0')` was removed from the tests.
Perhaps these tests were skipped before because Jenkins used numpy 1.9?

**Comment by niboshi** [comment]
I tested locally with master branch.
I couldn't replicate the error with `TestRfftn` but `TestRfft` caused similar errors.

**Comment by niboshi** [comment]
I couldn't replicate `TestRfftn` error in this PR branch, either.

**Comment by emcastillo** [comment]
It is strange that they only happen with python 3.5 ...
Other python versions are tested with newer numpy versions, so they are executed ...

**Comment by niboshi** [comment]
I sent a PR to fix `TestRfft` by adjusting tolerance (#3056), but I don't think the above `TestRfftn` error  can be fixed similarly (numerical errors are too large).

Btw, could you retrigger CI to see if the error is persistent?

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
@peterbell10 @grlee77 Do you guys have clues on why `rfftn` tests are failing here?

**Comment by grlee77** [comment]
I think this may be CUDA bug mentioned in https://github.com/cupy/cupy/issues/2569#issuecomment-571863448
It looks like those failing cases were using a P100 (Pascal architecture) GPU.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**